src/OrleansRuntime/GrainTypeManager/GrainTypeManager.cs

```diff
@@ -271,7 +271,11 @@ public IGrainMethodInvoker GetInvoker(string genericGrainType = null)
                 var typeArgs = TypeUtils.GenericTypeArgs(genericGrainType);
                 var concreteType = baseInvokerType.MakeGenericType(typeArgs);
                 var inv = (IGrainMethodInvoker) Activator.CreateInstance(concreteType);
-                cachedGenericInvokers[genericGrainType] = inv;
+                lock (cachedGenericInvokers)
```

Comment:
That lock is not enough. You also need to lock on the read access `if (cachedGenericInvokers.ContainsKey(genericGrainType))`.
Also, we usually lock on `private readonly object lockable` instead of arbitrary member fields. It is a more robust pattern for future modifications (member field becomes public, need to lock on more than one field, ...)